packages/alfa-rules/src/common/predicate/has-non-wrapped-text.ts

```diff
@@ -0,0 +1,42 @@
+import { Cache } from "@siteimprove/alfa-cache";
+import { Device } from "@siteimprove/alfa-device";
+import { Element, Text } from "@siteimprove/alfa-dom";
+import { Predicate } from "@siteimprove/alfa-predicate";
+import { Refinement } from "@siteimprove/alfa-refinement";
+
+import { hasChild } from "./has-child";
+import { hasComputedStyle } from "./has-computed-style";
+
+const { and, or, test } = Refinement;
+const { isElement } = Element;
+const { isText } = Text;
+
+const cache = Cache.empty<Device, Cache<Element | Text, boolean>>();
+
+export function hasNonWrappedText(device: Device): Predicate<Element | Text> {
+  return function hasNonWrappedText(node): boolean {
+    return cache.get(device, Cache.empty).get(
+      node,
+      () =>
+        test(
+          and(
+            hasChild(isText, {
+              flattened: true,
+            }),
+            hasComputedStyle(
+              "white-space",
+              (whiteSpace) => whiteSpace.value === "nowrap",
+              device
+            )
+          ),
+          node
+        ) ||
+        node
+          .children({
+            flattened: true,
+          })
+          .filter(or(isElement, isText))
+          .some(hasNonWrappedText)
```

Comment:
Side tracking… 🤔 
We do have quite a bunch of predicates that check some atomic stuff for a node and (recursively) either its ancestors (the full `isVisible` family) or its descendant (as here).
Wondering whether we could have a wrapper in `alfa-dom` like `someAncestor(predicate: Predicate<Node>): Predicate<Node>)` (or `everyAncestor`, `someInclusiveDescendant`, …) Might get tricky to handle caching on these.
Not sure it's worth the effort… Definitely not a topic for this PR.

Reply:
I feel that such a method mixes two concerns; ancestor traversal and predicate application on a sequence. To me, it seems like the composition `.ancestors(traversalOptions).<some | every | all>(predicate)` is favourable to `.<some | every | all>Ancestors(predicate, traversalOptions)`, especially since the latter requires three additional methods for something that can already be achieved with the four existing methods.

Reply:
You are perfectly right 😄 
Then I guess the question is why we have so many recursive predicates instead of using `.ancestors().some(predicate)` 😅

Reply:
Caching 🙈

Reply:
🤔 
```javascript
const cache = Cache.empty<Node, boolean>();

function foo(node: Node): boolean {
  return cache.get(node, () => !isElement(node) || node.id.isSome()
}

function bar(node: Node): boolean {
  return foo.inclusiveAncestors().some(foo);
}
```
should handle the cache and separate the actual test from the DOM tree walk 🤔 
`foo` could even be inlined if its test doesn't make sense in isolation (e.g. `isRendered` doesn't make sense if not testing ancestors).

Reply:
That only caches the result of each invocation of `foo()`, but you still potentially walk far up the inclusive ancestor chain every time `bar()` is called. We want to cache the entire recursive computation, making each step along the inclusive ancestor chain reusable.